### DISCUSSION TITLE
뿌튀인가 뭔가 그거때문에 스트레스에여

### DISCUSSION THREAD

**Discussion by yje0404N** [storycraft/node-kakao#644] [General]
요즘 소위 뿌튀라고해서 해외계정으로 한번에 5~10명씩 접속해서 
광고도배를하는데 패킷이 한번에 빠르게 몰려와서 그런가
강퇴가 안되네요 로그를 보면 Session is not created 이렇게 뜨는데..

오류 발생시 재로그인하게는 했는데
문제는 폭탄테러네요.. 폭탄테러가 오면 무조건 세션이 끊기니..

hidechat이랑 kickuser가 1명이 1~2개하면 막아지긴하는데

이런 폭탄테러의 경우 막을 방법이 없나요??
광고때문에 머리가 어질어질하네요

**Comment by ZZLOCO** [comment]
딜레이를 주어야 합니다. 
패킷을 빠르게 여러개 전송을 하는경우 그럴 수 있습니다.

**Reply by yje0404N** [reply]
딜레이를 어떤식으로 줘야하나요?
function Sleep (delay) {
  var start = new Date().getTime();
  while (new Date().getTime() < start + delay);
}

 channel.kickUser(data.chat.sender);
Sleep (100);

이렇게 써도 똑같은 현상이라..

**Reply by MoonJR** [reply]
@yje0404N 아마 저렇게 주시면 퇴장 밑으로는 딜레이가 들어가도 메시지가 도착하는건 비동기로 도착하기때문에 딜레이의 역활이 잘 안될것 같네요(모든 코드를 봐야겠지만요)

깔끔히 하려면 강퇴할 사용자를 큐에 넣었다가 0.1초마다 꺼내서 하나씩 지워줘야 비동기 처리가 쉬울것같네요
(0.1초로 해봤지만 0.1초도 죽더라구요 0.3~4초정도가 적당하지 않나 싶네요)

**Reply by yje0404N** [reply]
> @yje0404N 아마 저렇게 주시면 퇴장 밑으로는 딜레이가 들어가도 메시지가 도착하는건 비동기로 도착하기때문에 딜레이의 역활이 잘 안될것 같네요(모든 코드를 봐야겠지만요)
> 
> 깔끔히 하려면 강퇴할 사용자를 큐에 넣었다가 0.1초마다 꺼내서 하나씩 지워줘야 비동기 처리가 쉬울것같네요
> (0.1초로 해봤지만 0.1초도 죽더라구요 0.3~4초정도가 적당하지 않나 싶네요)

답변 감사드립니다..! 
sender를 큐에 넣고 꺼낸다는게 어떤 말씀이신지 제가 이해를 잘 못하겠습니다.

setTimeout(function(){
            channel.kickUser(data.chat.sender);
          },400);

이렇게 셋타임아웃을 이용하면 될까요??

**Reply by MoonJR** [reply]
@yje0404N 

음 논블럭 io에 대해 약간 이해가 필요한데 

위와 같이 작성이 실제로 폭탄으로 100개의 메시지가 들어왔을때 0.4초뒤에 100개를 한번에 지우는 꼴이 되어 죽게됩니다.

저같은 경우는

```
let messageQueue = [{
    channel:채널, chat: 메시지 
}]; // 그냥 예시라 보시면 됩니다.

talkClient.on('chat', (data, channel) => {
        //가려야 한다면
        messageQueue.push({channel:channel,chat:data.chat})
    }
)

function delay(ms) {
    return new Promise(resolve => setTimeout(() => {
        resolve();
    }, ms));
}

async function hideChat(){
    while(true){
        if (hideQueue.length > 0) {
               const hideData = hideQueue.shift();
               await hideData.channel.hideChat(hideData.chat);
        }
        await delay(300);
    }
}

hideChat().then();
```
음 예외처리등은 해야하지만 간단히 설명하면 요론 코드입니다. 메시지 수신시 큐에 쭉 담아두시고 300ms마다 큐를 확인해서 하나씩 가리게 하는것 입니다. 

저는 여기에 중복가리기 처리랑 등등 넣어서 사용하니 좀 괜찮더라구요

**Reply by ggoek34** [reply]
moonjr님 

중복가리기 처리가 뭘 뜻하는걸까요? 저는 폭탄오면 자꾸 찌꺼기들이 남아 있으니 죽겠네요 ㅠㅠ

**Reply by kiwioverflow** [reply]
중복 가리기 할 필요 없이 status 코드가 0이 아닐경우에만 다시 계속 가리기 요청 하는쪽으로 처리하시면 됩니다. 
요즘에는 가리기 요청이 누락되는 경우가 거의 없습니다.

**Reply by kiwioverflow** [reply]
가리기와 내보내기의 경우, 계정 상태에 따라 Rate Limit이 다 다릅니다.

금방 생성한 해외 계정의 경우 일반적으로 가리기는 초당 10개, 내보내기는 초당 약 3명정도로 제한이 되어있고,
국내 계정의 경우 계정마다 천차만별 이지만 실사용 계정이거나 오래된 계정이면 거의 제한이 없다 싶이 합니다. 

Rate Limit은 계정 활동이 많아지거나, 비정상적인 행동이 없을경우 완화되는 방식입니다. 
가리기의 경우 Rate Limit에 걸릴시 -303 응답코드를 리턴하며, 내보내기는 -810 응답코드를 리턴합니다.

참고 되시길 바랍니다.

**Reply by holiday22** [reply]
> 중복 가리기 할 필요 없이 status 코드가 0이 아닐경우에만 다시 계속 가리기 요청 하는쪽으로 처리하시면 됩니다. 요즘에는 가리기 요청이 누락되는 경우가 거의 없습니다.

status 코드가 0이 아닐경우를 어떻게 확인할 수 있을까요?

status 코드가 0이 아닐경우를 어떻게 확인할 수 있을까요?

data.chat에 있는 요소들 중에는 가려짐, 안가려짐을 확인할 수 있는 status 코드가 안 보여서요.

**Reply by kiwioverflow** [reply]
> > 중복 가리기 할 필요 없이 status 코드가 0이 아닐경우에만 다시 계속 가리기 요청 하는쪽으로 처리하시면 됩니다. 요즘에는 가리기 요청이 누락되는 경우가 거의 없습니다.
>
> @holiday22 
> status 코드가 0이 아닐경우를 어떻게 확인할 수 있을까요?
> 
> status 코드가 0이 아닐경우를 어떻게 확인할 수 있을까요?
> 
> data.chat에 있는 요소들 중에는 가려짐, 안가려짐을 확인할 수 있는 status 코드가 안 보여서요.

```
var result = await hideData.channel.hideChat(hideData.chat)
if(result.status==-303){
  //다시 큐 뒤로 넣는 코드
}
```
대충 저런식으로 하시면 됩니다.

**Reply by holiday22** [reply]
> > > 중복 가리기 할 필요 없이 status 코드가 0이 아닐경우에만 다시 계속 가리기 요청 하는쪽으로 처리하시면 됩니다. 요즘에는

정말 감사합니다.

**Reply by ggoek34** [reply]
계속 가리기 요청하는 쪽으로 처리했는데도 찌꺼기가 남는데 어떻게 해야 할까요? 
2초당 30개정도 보내버리네요 ㅠㅠ

**Reply by kiwioverflow** [reply]
> 계속 가리기 요청하는 쪽으로 처리했는데도 찌꺼기가 남는데 어떻게 해야 할까요? 2초당 30개정도 보내버리네요 ㅠㅠ

@ggoek34 가리기 요청에 리턴되는 Response 내용을 한번 확인해보셔야 할듯합니다.

일반적으로는 초당 최대 10개까지 가리기 요청을 보낼 수 있고, 이 이상으로는 -303 응답이 오거나 카카오가 고의적으로 응답을 느리게 주는 문제가 있습니다.

https://github.com/storycraft/node-kakao/blob/2c5ac555b5b9806bce49b4f00e280e941536da40/src/request/index.ts#L19

-303 응답이 돌아올 경우 큐 뒤로 넣어 다시 가리기 요청을 보내셔야 합니다.

그리고 2초당 30개면 심심찮게 영구 정지를 트리거하는 경우도 있으니, 가리기 초당 8~10회로 요청 횟수를 제한 하시는 게 좋습니다.

**Reply by ggoek34** [reply]
PARKKW472님 친절한 답변감사드립니다.  말씀하신대로 해보겠습니다

**Reply by RoRoGUMI** [reply]
> @yje0404N
> 
> 음 논블럭 io에 대해 약간 이해가 필요한데
> 
> 위와 같이 작성이 실제로 폭탄으로 100개의 메시지가 들어왔을때 0.4초뒤에 100개를 한번에 지우는 꼴이 되어 죽게됩니다.
> 
> 저같은 경우는
> 
> ```
> let messageQueue = [{
>     channel:채널, chat: 메시지 
> }]; // 그냥 예시라 보시면 됩니다.
> 
> talkClient.on('chat', (data, channel) => {
>         //가려야 한다면
>         messageQueue.push({channel:channel,chat:data.chat})
>     }
> )
> 
> function delay(ms) {
>     return new Promise(resolve => setTimeout(() => {
>         resolve();
>     }, ms));
> }
> 
> async function hideChat(){
>     while(true){
>         if (hideQueue.length > 0) {
>                const hideData = hideQueue.shift();
>                await hideData.channel.hideChat(hideData.chat);
>         }
>         await delay(300);
>     }
> }
> 
> hideChat().then();
> ```
> 
> 음 예외처리등은 해야하지만 간단히 설명하면 요론 코드입니다. 메시지 수신시 큐에 쭉 담아두시고 300ms마다 큐를 확인해서 하나씩 가리게 하는것 입니다.
> 
> 저는 여기에 중복가리기 처리랑 등등 넣어서 사용하니 좀 괜찮더라구요

조금만 더 상세히 설명해주실 분 없을까요? 예제는 없을련지..
